akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java

```diff
@@ -145,15 +145,18 @@ protected void postHttpBindingFailure(Throwable cause) {
    * Hook that lets the user specify the future that will signal the shutdown of the server whenever completed.
    */
   protected CompletionStage<Done> waitForShutdownSignal (ActorSystem system) {
-    return CompletableFuture.supplyAsync( () -> {
+    final CompletableFuture<Done> promise = new CompletableFuture<>();
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> promise.complete(Done.getInstance())));
+    CompletableFuture.runAsync(() -> {
       System.out.println("Press RETURN to stop...");
       try {
-        System.in.read();
+        if (System.in.read() >= 0)
```

Comment:
Do you really need the if in this case? `System.in.read` is blocking anyway and no matter what it returns we want the promise to complete.